lib/system.nim

```diff
@@ -2969,5 +2969,4 @@ proc arrayWith*[T](y: T, size: static int): array[size, T] {.raises: [].} =
     when nimvm:
       result[i] = y
     else:
-      # TODO: fixme it should be `=dup`
-      result[i] = y
+      result[i] = `=dup`(y)
```

Comment:
So, if I understand correctly, the current implementation of `arrayWith` is correct as it has been so far (as we actually *will* copy `y`, but in the example type we forbid copying values, so semantics seem correct to me). But maybe there would need to be an alternative implementation of this function to implement array initialization using the `default` proc to generate the default value instead of passing in a default value?

If my assumptions are correct, I guess I would need to make changes in the `defaultNodeField` proc. Somewhere around here: https://github.com/nim-lang/Nim/blob/15271dba2f4772a193f027d36c05b838dfff5bd4/compiler/sem.nim#L674 correct?

Reply:
Maybe instead of a generalized `arrayWith` the compiler could use an `arrayWithDefault` that does `result[i] = default(T)`, if `default(T)` behaves as expected in this case.

Reply:
Okay, I pushed an according change. I am not sure if my changes in `compiler/sem.nim` are sound, please check.